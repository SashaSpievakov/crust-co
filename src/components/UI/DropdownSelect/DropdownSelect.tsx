import { FC, KeyboardEvent, useEffect, useRef, useState } from 'react';

import {
  ArrowDown,
  ArrowUp,
  Label,
  Li,
  Popup,
  Section,
} from './DropdownSelect.styled';

interface DropdownSelectProps {
  chosenValue: string;
  sortNamesArr: string[];
  onSelect: (index: number) => void;
}

export const DropdownSelect: FC<DropdownSelectProps> = ({
  chosenValue,
  sortNamesArr,
  onSelect,
}) => {
  const [open, setOpen] = useState<boolean>(false);
  const sortRef = useRef<HTMLDivElement | null>(null);

  const handleOptionClick = (index: number) => {
    onSelect(index);
    setOpen(false);
  };

  const handleOptionKeyDownClick = (
    e: KeyboardEvent<HTMLElement>,
    i: number,
  ) => {
    if (e.code === 'Enter') handleOptionClick(i);
  };

  const handleLabelKeyDownClick = (e: KeyboardEvent<HTMLDivElement>) => {
    if (e.code === 'Enter') setOpen(!open);
  };

  useEffect(() => {
    const handleBodyClick = (e: MouseEvent) => {
      if (sortRef.current && !e.composedPath().includes(sortRef.current)) {
        setOpen(false);
      }
    };
    if (open) {
      document.body.addEventListener('click', handleBodyClick);
    }

    return () => {
      document.body.removeEventListener('click', handleBodyClick);
    };
  }, [open]);

  return (
    <Section ref={sortRef}>
      <Label
        onClick={() => setOpen(!open)}
        onKeyDown={(e) => handleLabelKeyDownClick(e)}
        tabIndex={0}
      >
        {open ? (
          <ArrowUp data-testid="popupIcon" />
        ) : (
          <ArrowDown data-testid="popupIcon" />
        )}
        <b>Sort by</b>
        <span data-testid="currenChosenSortName">{chosenValue}</span>
      </Label>
      <Popup open={open} aria-hidden={!open}>
        <ul aria-hidden={!open} data-testid="popupList">
          {sortNamesArr.map((sortName, i: number) => (
            <Li
              key={sortName}
              onClick={() => handleOptionClick(i)}
              onKeyDown={(e) => handleOptionKeyDownClick(e, i)}
              tabIndex={0}
              chosen={chosenValue === sortName}
            >
              {sortName}
            </Li>
          ))}
        </ul>
      </Popup>
    </Section>
  );
};
