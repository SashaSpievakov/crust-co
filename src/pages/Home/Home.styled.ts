import styled from 'styled-components';

const Title = styled.h2`
  margin: 35px 0;

  @media screen and (max-width: 650px) {
    margin: 20px 0;
  }
`;

const Top = styled.section`
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 30px;

  @media screen and (max-width: 900px) {
    flex-direction: column;
    align-items: start;
    margin-bottom: 15px;
  }
`;

const Block = styled(Top)`
  flex-direction: row;
  align-items: center;

  @media screen and (max-width: 900px) {
    margin-bottom: 50px;
  }

  @media screen and (max-width: 650px) {
    flex-direction: column;
    align-items: start;
  }
`;

const ErrorHeading = styled.h2`
  font-size: 32px;
  font-weight: 700;
  margin-bottom: 10px;

  @media screen and (max-width: 450px) {
    font-size: 25px;
  }
`;

const ErrorParagraph = styled.p`
  font-size: 22px;

  @media screen and (max-width: 650px) {
    font-size: 18px;
  }

  @media screen and (max-width: 450px) {
    font-size: 16px;
  }
`;

export { Title, Top, Block, ErrorHeading, ErrorParagraph };
