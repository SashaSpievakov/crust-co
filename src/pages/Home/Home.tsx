import { useEffect, useState } from 'react';

import { Title, Block, Top, Error } from '../../styles/Base.styled';
import { selectSort } from '../../store/slices/sortSlice';
import { selectCategory } from '../../store/slices/categorySlice';
import Categories from '../../components/Categories/Categories';
import Sort from '../../components/Sort/Sort';
import SearchItems from '../../components/SearchItems/SearchItems';
import { IPizzaItem } from '../../models/IPizzaItem';
import { useAppSelector } from '../../hooks/reduxHooks';
import pizzasAPI from '../../services/PizzasService';
import ProductsContainer from '../../components/ProductsContainer/ProductsContainer';

export const sortNamesArr: string[] = ['rating', 'price', 'A to Z'];

const Home = () => {
  const [pizzas, setPizzas] = useState<IPizzaItem[]>([]);
  const activeCategory = useAppSelector(selectCategory);
  const activeSort = useAppSelector(selectSort);

  const sortedActiveName = sortNamesArr[activeSort];

  const sortPropertyName = (property: string) => {
    if (property === 'A to Z') {
      return 'name&order=asc';
    }
    return `${property}&order=desc`;
  };

  const sortedPropertyName = sortPropertyName(sortedActiveName);

  const { data, isSuccess, isError, isLoading } = pizzasAPI.useFetchPizzasQuery(
    {
      activeCategory,
      sortedPropertyName,
    },
  );

  useEffect(() => {
    if (isSuccess) {
      setPizzas(data);
    }
  }, [activeCategory, sortedPropertyName, data, isSuccess]);

  return (
    <>
      <Top>
        <Categories />
        <Sort sortNamesArr={sortNamesArr} />
      </Top>
      <Block>
        <Title>All Pizzas</Title>
        <SearchItems />
      </Block>
      {isError ? (
        <Error>
          <h2>Request Error</h2>
          <p>Coudn&apos;t get store items. Try your request again later.</p>
        </Error>
      ) : (
        <ProductsContainer isLoading={isLoading} items={pizzas} />
      )}
    </>
  );
};

export default Home;
