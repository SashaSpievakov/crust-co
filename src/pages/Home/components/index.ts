export { Categories } from './Categories';
export { ProductsContainer } from './ProductsContainer';
