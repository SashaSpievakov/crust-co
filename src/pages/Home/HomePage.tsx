import { ChangeEvent, FC, useEffect, useState } from 'react';
import { useSearchParams } from 'react-router-dom';
import { ErrorRequest } from 'src/components';
import { useDebounce } from 'src/hooks';
import { Container } from 'src/styles/Base.styled';

import { DropdownSelect, SearchItems } from '../../components/UI';
import { IPizzaItem } from '../../models';
import { pizzasAPI } from '../../services';
import { modifySearchParamsName } from '../../utils/modifySearchParamsName';
import { Categories, ProductsContainer } from './components';
import { Block, Title, Top } from './HomePage.styled';

export const sortNamesArr: string[] = ['rating', 'price', 'A to Z'];

export const HomePage: FC = () => {
  const [searchParams, setSearchParams] = useSearchParams();
  const [pizzas, setPizzas] = useState<IPizzaItem[]>([]);
  const [searchValue, setSearchValue] = useState<string>('');
  const debouncedSearchValue = useDebounce(searchValue, 100);
  const [activeSort, setActiveSort] = useState<number>(
    parseInt(searchParams.get('sortBy') || '0', 10) || 0,
  );
  const [activeCategory, setActiveCategory] = useState<number>(
    parseInt(searchParams.get('category') || '0', 10) || 0,
  );

  const sortSearchParam = modifySearchParamsName(sortNamesArr[activeSort]);
  const { data, isSuccess, isError, isLoading } = pizzasAPI.useFetchPizzasQuery(
    {
      activeCategory,
      sortSearchParam,
    },
  );

  useEffect(() => {
    document.title = 'Crust & Co.';
  }, []);

  useEffect(() => {
    if (isSuccess) {
      setPizzas(data);
    }
  }, [activeCategory, sortSearchParam, data, isSuccess]);

  const handleSortSelect = (index: number): void => {
    setActiveSort(index);
    setSearchParams({
      sortBy: index.toString(),
      category: activeCategory.toString(),
    });
  };

  const handleCategorySelect = (index: number): void => {
    setActiveCategory(index);
    setSearchParams({
      category: index.toString(),
      sortBy: activeSort.toString(),
    });
  };

  const onChangeSearch = (e: ChangeEvent<HTMLInputElement>): void => {
    setSearchValue(e.target.value);
  };

  const onClickClose = () => {
    setSearchValue('');
  };

  return isError ? (
    <ErrorRequest />
  ) : (
    <Container data-testid="homePage">
      <Top>
        <Categories
          activeCategory={activeCategory}
          onSelect={handleCategorySelect}
        />
        <DropdownSelect
          chosenValue={sortNamesArr[activeSort]}
          sortNamesArr={sortNamesArr}
          onSelect={handleSortSelect}
        />
      </Top>
      <Block>
        <Title>All Pizzas</Title>
        <SearchItems
          value={searchValue}
          width="240px"
          onChange={onChangeSearch}
          onClickClose={onClickClose}
          expandedWidth="300px"
        />
      </Block>
      <ProductsContainer
        isLoading={isLoading}
        items={pizzas}
        searchValue={debouncedSearchValue}
      />
    </Container>
  );
};
