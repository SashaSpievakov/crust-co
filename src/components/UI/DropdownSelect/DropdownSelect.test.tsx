import '@testing-library/jest-dom';

import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { sortNamesArr } from '@src/pages/Home/HomePage';
import {
  rendererWithProviders,
  renderWithProvidersAndRoutes,
} from '@src/tests/helpers';

import { DropdownSelect } from './DropdownSelect';

describe('DropdownSelect Test', () => {
  test('should render the DropdownSelect component', () => {
    const snapshot = rendererWithProviders(
      <DropdownSelect
        sortNamesArr={sortNamesArr}
        chosenValue="rating"
        onSelect={() => null}
      />,
    );
    expect(snapshot).toMatchSnapshot();
  });

  describe('DropdownSelect general functionality', () => {
    beforeEach(() => {
      renderWithProvidersAndRoutes(
        <DropdownSelect
          sortNamesArr={sortNamesArr}
          chosenValue="rating"
          onSelect={() => null}
        />,
      );
    });

    test('should open the dropdown select', () => {
      const arrow = screen.getByTestId('popupIcon');
      const list = screen.getByRole('list', { hidden: true });

      userEvent.click(arrow);

      expect(list).toBeVisible();
    });

    test('should close the dropdown select on icon click', () => {
      const arrow = screen.getByTestId('popupIcon');
      const list = screen.getByRole('list', { hidden: true });

      userEvent.click(arrow);
      userEvent.click(screen.getByTestId('popupIcon'));

      expect(list).not.toBeVisible();
    });

    test('should close the dropdown select on list item click', () => {
      const sortCaption = screen.getByText(/sort by/i);
      const list = screen.getByRole('list', { hidden: true });

      userEvent.click(sortCaption);
      userEvent.click(screen.getByText(/price/i));

      expect(list).not.toBeVisible();
    });
  });

  describe('DropdownSelect in an application environment', () => {
    test('should close popup on an outside click', () => {
      renderWithProvidersAndRoutes(null, true);
      const sortCaption = screen.getByText(/sort by/i);
      const bodyHeading = screen.getByRole('heading', { level: 2 });
      const list = screen.getByTestId('popupList');

      userEvent.click(sortCaption);
      userEvent.click(bodyHeading);

      expect(list).not.toBeVisible();
    });
  });
});
