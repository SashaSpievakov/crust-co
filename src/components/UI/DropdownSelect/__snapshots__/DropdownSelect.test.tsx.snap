// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sort Test renders the Sort component 1`] = `
Object {
  "children": Array [
    <div
      className="sc-idXgbr fYmPoy"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
    >
      <svg
        className="sc-hHTYSt cijjmx"
        data-testid="popupIcon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7.247 11.14 2.451 5.658C1.885 5.013 2.345 4 3.204 4h9.592a1 1 0 0 1 .753 1.659l-4.796 5.48a1 1 0 0 1-1.506 0z"
        />
      </svg>
      <b>
        Sort by
      </b>
      <span
        data-testid="currenChosenSortName"
      >
        rating
      </span>
    </div>,
    <section
      aria-hidden={true}
      className="sc-dmctIk ikiszO"
      open={false}
    >
      <ul
        aria-hidden={true}
        data-testid="popupList"
      >
        <li
          className="sc-kgTSHT kXJAEI"
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
        >
          rating
        </li>
        <li
          className="sc-kgTSHT azQnF"
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
        >
          price
        </li>
        <li
          className="sc-kgTSHT azQnF"
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
        >
          A to Z
        </li>
      </ul>
    </section>,
  ],
  "props": Object {
    "className": "sc-bjfHbI FKPCf",
  },
  "type": "section",
}
`;
